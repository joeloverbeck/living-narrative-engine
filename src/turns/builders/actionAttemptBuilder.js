// src/turns/builders/attemptActionBuilder.js
/**
 * @typedef {import('../../interfaces/coreServices.js').ILogger} ILogger
 */

/**
 * @typedef {Object} CoreAttemptActionPayload
 * @property {'core:attempt_action'} eventName
 * @property {string} actorId
 * @property {string} actionId
 * @property {Object.<string, any>} params
 */

/**
 * Builder for creating CoreAttemptActionPayloads from ActionComposite-like objects.
 * Pure aside from logging side-effects.
 */
export class AttemptActionBuilder {
  /**
   * @param {ILogger} logger
   */
  constructor(logger) {
    this.logger = logger;
  }

  /**
   * Build a core:attempt_action payload from an ActionComposite.
   *
   * @param {string} actorId
   * @param {Object} composite
   * @param {number} composite.index
   * @param {string} composite.actionId
   * @param {Object} composite.params
   * @param {string} composite.description
   * @returns {CoreAttemptActionPayload}
   */
  build(actorId, composite) {
    // Deep-copy params
    const paramsCopy = JSON.parse(JSON.stringify(composite.params));

    // Scoped logger
    const actionLogger = this.logger.child({ idx: composite.index });
    actionLogger.info(
      `actor=${actorId} idx=${composite.index} actionId=${composite.actionId} ` +
        `params=${JSON.stringify(paramsCopy)} description="${composite.description}"`
    );

    return {
      eventName: 'core:attempt_action',
      actorId,
      actionId: composite.actionId,
      params: paramsCopy,
    };
  }
}
