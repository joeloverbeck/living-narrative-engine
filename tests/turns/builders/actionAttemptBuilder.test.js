// tests/builders/attemptActionBuilder.spec.js

import { mock } from 'jest-mock-extended';
import { describe, beforeEach, it, expect } from '@jest/globals';
import { AttemptActionBuilder } from '../../../src/turns/builders/actionAttemptBuilder.js';

describe('AttemptActionBuilder', () => {
  let logger;
  let childLogger;
  let builder;

  const actorId = 'actor1';
  const composite = {
    index: 1,
    actionId: 'testAction',
    params: { foo: 'bar', nested: { a: 1 } },
    description: 'A test action',
  };

  beforeEach(() => {
    // 1) Create mocks
    logger = mock();
    childLogger = mock();

    // 2) Wire up child logger and spy on info()
    logger.child.mockReturnValue(childLogger);
    childLogger.info = jest.fn();

    // 3) Instantiate builder under test
    builder = new AttemptActionBuilder(logger);
  });

  it('build() output matches snapshot for a sample composite', () => {
    const result = builder.build(actorId, composite);
    expect(result).toMatchSnapshot();
  });

  it('mutation attempt on returned payload does not affect original composite', () => {
    const result = builder.build(actorId, composite);

    // Mutate the returned params
    result.params.foo = 'changed';
    result.params.nested.a = 2;

    // Original composite.params should remain untouched
    expect(composite.params.foo).toBe('bar');
    expect(composite.params.nested.a).toBe(1);
  });

  it('logger mock receives the expected formatted message', () => {
    const result = builder.build(actorId, composite);

    // Logger.child must be called with the index
    expect(logger.child).toHaveBeenCalledWith({ idx: composite.index });

    // Construct expected log string
    const expectedMessage =
      `actor=${actorId} ` +
      `idx=${composite.index} ` +
      `actionId=${composite.actionId} ` +
      `params=${JSON.stringify(result.params)} ` +
      `description="${composite.description}"`;

    // Verify info() got exactly that
    expect(childLogger.info).toHaveBeenCalledWith(expectedMessage);
  });
});
