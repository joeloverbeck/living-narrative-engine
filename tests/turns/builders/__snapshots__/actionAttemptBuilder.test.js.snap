// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttemptActionBuilder build() output matches snapshot for a sample composite 1`] = `
{
  "actionId": "testAction",
  "actorId": "actor1",
  "eventName": "core:attempt_action",
  "params": {
    "foo": "bar",
    "nested": {
      "a": 1,
    },
  },
}
`;
